crates/muvm/src/guest/net.rs

```diff
@@ -1,14 +1,215 @@
+fn do_dhcp(rtnl: &NlRouter) -> Result<()> {
+    // Temporary link-local address and route avoid the need for raw sockets
+    route4_eth0(rtnl, Rtm::Newroute, Ipv4Addr::UNSPECIFIED)?;
+    addr4_eth0(rtnl, Rtm::Newaddr, Ipv4Addr::new(169, 254, 1, 1), 16)?;
+
+    // Send request (DHCPDISCOVER)
+    let socket = UdpSocket::bind("0.0.0.0:68").expect("Failed to bind");
+    let mut buf = [0; 576 /* RFC 2131, Section 2 */ ];
+
+    const REQUEST: [u8; 300 /* From RFC 951: >= 60 B of options */ ] = [
+        1 /* REQUEST */, 0x1 /* Ethernet */, 6 /* hlen */, 0 /* Hops */,
+        1, 2, 3, 4 /* XID */, 0, 0 /* Seconds */, 0x80, 0x0 /* Flags */,
+        0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0, /* All-zero (four) addresses */
+        0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0, /* 16B HW address: who cares */
+        /* 32 bytes per row: 64B 'sname', plus 128B 'file' (RFC 1531) */
+        0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+        0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+        0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+        0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+        0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+        0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+        0x63, 0x82, 0x53, 0x63, /* DHCP (magic) cookie, then options: */
+        53, 1, 1 /* DISCOVER */, 80, 0 /* Rapid commit */, 0xff, // Done
+        0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+        0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0 /* 54B paaaadding */
+    ];
+
+    socket.set_broadcast(true)?;
+    socket.send_to(&REQUEST, "255.255.255.255:67")?;
+
+    // Keep IPv6-only fast
+    let _ = socket.set_read_timeout(Some(Duration::from_millis(100)));
+
+    // Get and process response (DHCPACK) if any
+    if let Ok((len, _)) = socket.recv_from(&mut buf) {
+        let msg = &mut buf[..len];
+
+        let addr = Ipv4Addr::new(msg[16], msg[17], msg[18], msg[19]);
+        let mut netmask = Ipv4Addr::UNSPECIFIED;
+        let mut router = Ipv4Addr::UNSPECIFIED;
+        let mut p: usize = 240;
+
+        while p < len {
+            let o = msg[p];
+            let mut l: u8;
+            l = msg[p + 1];
+
+            if o == 1 {           // Option 1: Subnet Mask
+                netmask = Ipv4Addr::new(msg[p + 2], msg[p + 3],
+                                        msg[p + 4], msg[p + 5]);
+            } else if o == 3 {    // Option 3: Router
+                router =  Ipv4Addr::new(msg[p + 2], msg[p + 3],
+                                        msg[p + 4], msg[p + 5]);
+            } else if o == 0xff { // Option 255: End (of options)
+                break;
+            }
+
+            l += 2; // Length doesn't include code and length field itself
+            p += l as usize;
+        }
+
+        let prefix_len : u8 = netmask.to_bits().leading_ones() as u8;
```

Comment:
I was wondering why the bump to Rust 1.80, and this is the reason...

https://doc.rust-lang.org/stable/core/net/struct.Ipv4Addr.html#method.to_bits

(No action necessary here.)

Reply:
Correct :)